Subsystems

ky28059 opened the pull request.

e3l committed:
Jetson camera wrapper

eggaskin committed:
added elevator and tank implementation

e3l committed:
Tank drive code, untested

Ayeeti committed:
add elevator config

eggaskin committed:
reviewed added code, default commands added

eggaskin committed:
elevator commands and controller setup

e3l committed:
Working tank and elevator subsystems

PinkBlueSky committed:
overhaul subsystem structure + overhaul config file

PinkBlueSky committed:
fix end method in dtcommand

PinkBlueSky committed:
improved handling of elevatorUpIsPositive const

PinkBlueSky committed:
change elevator command structure

PinkBlueSky committed:
put in config txt thing again

ky28059 committed:
Use `WPI_TalonSRX`, `DifferentialDrive`

ky28059 committed:
Use builtin `squareInput`

ky28059 committed:
Add third set of Talons

ky28059 committed:
Remove `DifferentialDrive`

ky28059 committed:
Revert "Add third set of Talons"

This reverts commit 1b5ea6d2a5522477e4151abe8c168437a8f0d1df.

PinkBlueSky committed:
add joystick code

PinkBlueSky committed:
Add joystick capability

PinkBlueSky committed:
Add controller switching

Ippyb committed:
code for testing gyro, ecoder, limit switch, and color+distance sensor. Supposed to print values from sensors. Need to import rev library.

Ippyb committed:
sensor testing code

PinkBlueSky committed:
editing jetson socket so it's using constants instead of bigdata

PinkBlueSky committed:
editing jetson socket so it's using constants instead of bigdata

PinkBlueSky committed:
add networktables

PinkBlueSky committed:
Add Jetson NT testing code

PinkBlueSky committed:
Update NetworkTables data to match vision code

PinkBlueSky committed:
Update NetworkTables data to match vision code

PinkBlueSky committed:
Clean up

PinkBlueSky committed:
Clean up

PinkBlueSky committed:
Add httpcamera for testing camera stream

ky28059 committed:
Recreate shooter branch from refactored Ethan code

ky28059 committed:
Skeleton `shoot()` method, some cleanup

ky28059 committed:
Add hood SparkMax

ky28059 committed:
Remember to add a constant :flushed:

PinkBlueSky committed:
Update Jetson stuff

ky28059 committed:
PID testing

ky28059 committed:
Jetson refactoring

ky28059 committed:
Add velocity closed loop testing code

Ippyb committed:
skeleton code for internal subsystem. Includes 2 color sensors, a talon (will replace w/ cim), a method that determines when to activate/deactivate motor based on color.

ky28059 committed:
Jetson restructuring, `periodic()` skeleton code

ky28059 committed:
Rename shooter to turret

Shooting has been delegated to internals!

ky28059 committed:
Merge commit part 2

ky28059 committed:
Internals refactoring

ky28059 committed:
Initiate InternalSubsystem in RobotContainer

ky28059 committed:
Add `turntableAligned()` for internals

Ippyb committed:
shooter request updates

Ippyb committed:
added a skeleton method for launching ball into shooter

ky28059 committed:
Change initialization to constructor arguments

ky28059 committed:
Skeleton climb code, cleanup

ky28059 committed:
Intake skeleton

ky28059 committed:
Use intake enum, add intake commands

ky28059 committed:
Reformat to 4 spaces

Finally, my master plan is complete :smiling_imp:

In all seriousness, all the new files being created in each branch were all using 4 space indents and the constant refactorings to preserve an unliked indent size inherited from 2 (3?) years ago was getting annoying. 4 space is the future!

ky28059 committed:
Initialize new subsystems in RobotContainer

ky28059 committed:
Fix tank spacing, better comments

ky28059 committed:
Restructure intake position, PID constant splitting

ky28059 committed:
Implement new climb design

Ippyb committed:
added motorport2 constant

Ippyb committed:
added a ballDetected method, updated fields to include another motor, undefined ball count method

ky28059 committed:
Intake / internals integration

Ippyb committed:
updateBallCount

ky28059 committed:
Ball count tracking code

Ippyb committed:
merge

ky28059 committed:
Jetson constants organization, better reject logic

ky28059 committed:
Set up shuffleboard for PID testing

ky28059 committed:
Fix internals motor flow

ky28059 committed:
Use enum for turret mode

Ippyb committed:
got rid of TODOs

ky28059 committed:
Roughly implement the climb plan

It's pretty complex.

ky28059 committed:
775 instead of NEO for turret hood

ky28059 committed:
Update internals with IR sensors

ky28059 committed:
Use command groups for climb

ky28059 committed:
Implement module states for turret

This is so that a) drivers know what the status of the shooter is and b) internals can reject quicker (on orange) instead of waiting for a perfectly lined up shot.

ky28059 committed:
Final-ish button bindings

ky28059 committed:
Final rebase commit

ky28059 committed:
`GRTSubsystem` changes, subsystem implementation

ky28059 committed:
Implement driver intake running

ky28059 committed:
Non-conflicting motor ports

ky28059 committed:
New constants, third tank NEO

ky28059 committed:
Turntable blind spot logic, soft limits

ky28059 committed:
Degree conversion constant stubs

ky28059 committed:
Remove mjpg prefix in jetson

PinkBlueSky committed:
Working vision :D

PinkBlueSky committed:
merge w/ shooter

ky28059 committed:
Temporary intake + DT testing code

The drivetrain and intake rollers work as expected, but the intake deploy was not moving the motor upwards (despite the percent output being set correctly). I'd advise whoever's testing this tomorrow to see if running the current code and pressing the A button causes the talon to begin blinking green (positive percent output); if it does, then it's a motor issue, otherwise its a code issue (which is weird, because `deploy.get()` seems to say that the deploy controller is indeed being set to the proper percent output).

ky28059 committed:
Replace exit with staging sensor

This should work because staging initially stops the ball the moment it detects it (when the top of the ball is at the sensor). While staging still detects the ball (for the entire diameter of the ball as it moves up) we keep spinning the top, and as long as the bottom of the ball is still seen by staging when the top of the ball reaches the turret this code should work. We can repurpose the `shotRequested = false` code which runs when the ball leaves internals to also handle ball count decrementation.

e3l committed:
Testing of subsystems, lowered periodic loop times, moved localization to thread

ky28059 commented on src/main/java/frc/robot/subsystems/tank/TankSubsystem.java:
```diff
@@ -1,31 +1,31 @@
+package frc.robot.subsystems.tank;
 
-import com.kauailabs.navx.frc.AHRS;
+import static frc.robot.Constants.TankConstants.bLeftMotorPort;
```
Why this instead of the previous 
```java
import static frc.robot.Constants.TankConstants.*;
```
?

e3l replied:
auto format

proan replied:
Importing with a `*` is a bad idea. You don't know all that you're bringing in to the namespace. Better to be explicit.

ky28059 commented on src/main/java/frc/robot/subsystems/tank/TankSubsystem.java:
```diff
@@ -1,31 +1,31 @@
-package frc.robot.subsystems;
+package frc.robot.subsystems.tank;
```
Subsystem nesting is back? 😩 
Can we keep the pose estimator thread in `/subsystems` and not in its own folder?

e3l replied:
i'd rather not have to read 6 file names to get to a subsystem, i guess it depends on the others

ky28059 commented on src/main/java/frc/robot/subsystems/tank/TankSubsystem.java:
```diff
@@ -185,57 +180,34 @@ public void setTankDriveVoltages(double leftVoltage, double rightVoltage) {
 
     /**
      * Gets the estimated current position of the robot.
+     * 
```
VSCode autoformat 😕

ky28059 replied:
The extra newlines in javadoc and comment wrapping actually makes it harder to read than properly-formatted human written code imo

proan replied:
Who is writing the formatting style VSCode is using? I think you can specify your own. Also lint.

ky28059 commented on src/main/java/frc/robot/subsystems/tank/PoseEstimatorThread.java:
```diff
@@ -0,0 +1,116 @@
+        private final RelativeEncoder leftEncoder;
+        private final RelativeEncoder rightEncoder;
+
+        private DifferentialDriveWheelSpeeds lastWheelSpeeds;
```
Why is this saved as a field? Is constructing the `DifferentialDriveWheelSpeeds` object really that expensive?

e3l replied:
No, but calling the encoders again is -- that's what I have been saying about minimizing shuffleboard calls, .set calls, .get calls... I/O takes time

proan replied:
In addition to @e3l 's comments:

Because you're essentially running an embedded system, you should think carefully about when objects are created and destroyed. Contrary to typical programming, it's common in embedded to have global variables just so the memory is all allocated at the beginning and there's little to no dynamic memory allocation.

Secondly, any operation like constructing an object has a processing penalty. In the embedded case, it is often better to do this all at startup rather than in your control loop. You want very consistent timing for your control loops, and thus you should prefer to keep the operations in those loops well known.

ky28059 commented on src/main/java/frc/robot/subsystems/tank/PoseEstimatorThread.java:
```diff
@@ -0,0 +1,116 @@
+
+        @Override
+        public void run() {
+            while (true) {
```
There's no `Thread.sleep()` in this loop.

e3l replied:
... yes, because we want to be estimating our pose as often as possible

proan replied:
Uhhhh. @e3l you probably should sleep once an a while, and you probably don't need to update _that_ often. 10 Hz should be plenty. Do we know the timing here?

e3l replied:
Haven't extensively profiled the performance of this thread or how accuracy changes with different update speeds.

We do know that it's more I/O blocked than anything else, which means that having no sleep doesn't mean it'll be expensive or slow down other threads, necessarily, right?

ky28059 committed:
Enforce style

ky28059 committed:
Better `ModuleState` names

ky28059 committed:
Jetson reformatting

ky28059 committed:
Fix internals current drawn method

e3l committed:
More subsystem testing and some refactoring

ky28059 committed:
Enforce style again.

ky28059 committed:
Attempt turret `r` and `theta` state system

PinkBlueSky committed:
Bare bones intake testing

e3l committed:
Internals color sensor thread

ky28059 committed:
Measure color constants, reduce IO waste

ky28059 committed:
More testing setup

e3l committed:
Upgrade to WPILib 2022.3.1

PinkBlueSky committed:
Update encoder value; comment out internals

ky28059 committed:
Button binding fix, use integrated encoders for DT

Also, experimental style changes! Unsure how I like the comma separated declaration, but it is certainly less wordy than repeating `private final JoystickButton` 8 times.

ky28059 committed:
Intake, internals, turret testing

- Commented out PowerController (it was causing power drops and stuttering, will test causes and try fixes later)
- Bind turntable and hood to mech controller joysticks
- Bind left trigger to internals, right to intake
- Preliminary internals restructuring to replace I2C onboard color sensor

ky28059 committed:
Boolean state for storage rejection checking

ky28059 committed:
More internals testing

ky28059 committed:
Internals testing part 2 -- dummy shoot method

It works half of the time! See slack for details about inconsistencies

ky28059 committed:
Shooter testing

- Bind mech y to turn on flywheel, mech b to turn off
- Manual override of internals motors on mech triggers

eggaskin reviewed (approved):
At this point we are using shooter like develop, so even though I haven't reviewed all this code I'm going to merge it. Let's actually use develop hooray
